pynest/nest/tests/test_connect_symmetric_pairwise_bernoulli.py

```diff
@@ -0,0 +1,149 @@
+                # degrees = self.comm.gather(degrees, root=0)
+                # if self.rank == 0:
+                if degrees is not None:
+                    chi, p = hf.chi_squared_check(degrees, expected, self.rule)
+                    pvalues.append(p)
+                hf.mpi_barrier()
+            if degrees is not None:
+                ks, p = scipy.stats.kstest(pvalues, 'uniform')
+                self.assertTrue(p > self.stat_dict['alpha2'])
+
+    def testAutapses(self):
+        conn_params = self.conn_dict.copy()
+        conn_params['autapses'] = True
+        N = 10
+
+        # test that autapses are not permitted
+        hf.nest.ResetKernel()
+        pop = hf.nest.Create('iaf_psc_alpha', N)
+        with self.assertRaises(hf.nest.NESTError):
+            hf.nest.Connect(pop, pop, conn_params)
+
+        # test that autapses were excluded
+        conn_params['p'] = 1. - 1. / N
+        conn_params['autapses'] = False
+        hf.nest.ResetKernel()
+        pop = hf.nest.Create('iaf_psc_alpha', N)
+        hf.nest.Connect(pop, pop, conn_params)
+        M = hf.get_connectivity_matrix(pop, pop)
+        hf.mpi_assert(np.diag(M), np.zeros(N), self)
+
+    def testMultapses(self):
+        conn_params = self.conn_dict.copy()
+        conn_params['multapses'] = False
+        N = 10
+
+        # test that multapses must be permitted
+        hf.nest.ResetKernel()
+        pop = hf.nest.Create('iaf_psc_alpha', N)
+        with self.assertRaises(hf.nest.NESTError):
+            hf.nest.Connect(pop, pop, conn_params)
+
+        # test that multapses can only arise from symmetric
+        # connectivity
+        conn_params['p'] = 1. - 1. / N
+        conn_params['multapses'] = True
+        hf.nest.ResetKernel()
+        pop = hf.nest.Create('iaf_psc_alpha', N)
+        hf.nest.Connect(pop, pop, conn_params)
+
+        conn_dict = collections.defaultdict(int)
+        for conn in hf.nest.GetConnections():
+            key = tuple(conn[:2])
+            conn_dict[key] += 1
+            self.assertTrue(conn_dict[key] <= 2)
+
+    def testMakeSymmetric(self):
+        conn_params = self.conn_dict.copy()
+        N = 100
+
+        # test that make_symmetric must be enabled
+        conn_params['make_symmetric'] = False
+        hf.nest.ResetKernel()
+        pop = hf.nest.Create('iaf_psc_alpha', N)
+        with self.assertRaises(hf.nest.NESTError):
+            hf.nest.Connect(pop, pop, conn_params)
+
+        # test that all connections are symmetric
+        conn_params['make_symmetric'] = True
+        hf.nest.ResetKernel()
+        pop = hf.nest.Create('iaf_psc_alpha', N)
+        hf.nest.Connect(pop, pop, conn_params)
+
+        conns = set()
+        for conn in hf.nest.GetConnections():
+            key = tuple(conn[:2])
+            conns.add(key)
+
+        for conn in hf.nest.GetConnections():
+            key = tuple(conn[:2])
+            self.assertTrue(key[::-1] in conns)
+
+
```

Comment:
I am missing an test that actually checks if the all created connections are symmetric. Something like

```py
import nest
import numpy as np
import test_connect_helpers as hf
from mpi4py import MPI

n = nest.Create('hh_psc_alpha_gap',40)

rule = 'symmetric_pairwise_bernoulli'
p = 0.5
conn_dict = {'rule': rule, 'p': p, 'multapses': True, 'autapses': False, 'make_symmetric': True}

nest.Connect(n ,n , conn_dict, {'model':'gap_junction'})

local_matrix = hf.get_connectivity_matrix(n,n)

global_matrix = hf.gather_data(local_matrix)

if MPI.COMM_WORLD.Get_rank() == 0:
  #print global_matrix
  assert(np.allclose(global_matrix, global_matrix.T, atol=1e-8))
```
which is called from e.g. 1,4 and 8 MPI processes. Is there an easy way to include this?

Reply:
Actually, we were testing for symmetric connections in line 128--135.
1) add all pairs of source gid, target gid to a set
2) check whether the reversed pairs are in the set

Shouldn't this do the trick?
```python
        conns = set()
        for conn in hf.nest.GetConnections():
            key = tuple(conn[:2])
            conns.add(key)

        for conn in hf.nest.GetConnections():
            key = tuple(conn[:2])
            self.assertTrue(key[::-1] in conns)
```

I agree that we do not currently test for the distributed case. As far as I can see, this is done in `test_connect_all_patterns.py`. However this test fails on my system. Will create an issue for that and include MPI tests as soon as I've figured out what goes wrong.

Reply:
@jakobj Oh sorry, of course you are right, I somehow missed that part and thought that `testMakeSymmetric(self)` only checks that `make_symmetric` must be enabled. I think the current test is sufficient to test the functionality without MPI (also in theory there could be several identical connections in one direction and only one connection in the opposite direction and the test would not detect it).

Once an MPI test is added I am all happy :+1: